src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py

```diff
@@ -24,46 +67,166 @@ def generate_itin(
+        if counter >= 10:
+            logger.info(
+                f"Resampled ITIN {counter} times and data still contains {duplicate_mask.sum()}"
+                f"duplicates remaining.  Check data or usage of randomness stream."
+            )
+            break
+        new_additional_key = f"{additional_key}_{counter}"
+        itins.loc[duplicate_mask] = generate_itin(duplicate_mask.sum(), new_additional_key)
+        duplicate_mask = itins.duplicated()
+        counter += 1
+    return itins.to_numpy()
+
+
+def generate_ssns(
+    size: int,
+    additional_key: Any,
+    randomness: RandomnessStream,
+) -> np.ndarray:
+    """
+    Generate a np.ndarray of length `size` of unique SSN values.
+
+        first three digits 1 and 899 (inclusive), 666 disallowed, followed by dash
+        next two digits between 1-99 (inclusive), followed by a dash
+        last four digits between 1-9999 (inclusive)
+
+    :param size: The number of itins to generate
+    :param additional_key: Additional key to be used by randomness
+    :param randomness: RandomnessStream stream to use
+
+    """
+
+    def generate_ssn(
+        count: int,
+        extra_additional_key: Any,
+    ) -> pd.DataFrame:
+        ssn = pd.DataFrame(index=pd.Index(range(count)))
 
-    # Rescale the group digits to match expected ranges
-    pre_shift_groups = itin["group"].copy()
-    itin.loc[pre_shift_groups <= 16, "group"] += 49
-    itin.loc[((pre_shift_groups <= 35) & (pre_shift_groups >= 17)), "group"] += 53
-    itin.loc[((pre_shift_groups <= 38) & (pre_shift_groups >= 36)), "group"] += 54
-    itin.loc[pre_shift_groups >= 39, "group"] += 55
-
-    # Serial numbers 1-9999 inclusive
-    itin["serial"] = random_integers(
-        min_val=1,
-        max_val=9999,
-        index=itin.index,
-        randomness=randomness,
-        additional_key=f"{additional_key}_itin_serial",
+        area = random_integers(
+            min_val=1,
+            max_val=899,
+            index=ssn.index,
+            randomness=randomness,
+            additional_key=f"{extra_additional_key}_ssn_area",
+        )
+        area = np.where(area == 666, 667, area)  # XXX is this reasonable?
+        ssn["ssn_area"] = area
+
+        group = random_integers(
+            min_val=1,
+            max_val=99,
+            index=ssn.index,
+            randomness=randomness,
+            additional_key=f"{extra_additional_key}_ssn_group",
+        )
+        ssn["ssn_group"] = group
+
+        serial = random_integers(
+            min_val=1,
+            max_val=9999,
+            index=ssn.index,
+            randomness=randomness,
+            additional_key=f"{extra_additional_key}_ssn_serial",
+        )
+        ssn["ssn_serial"] = serial
+
+        ssn["ssn"] = ssn.ssn_area.astype(str).str.zfill(3) + "-"
+        ssn["ssn"] += ssn.ssn_group.astype(str).str.zfill(2) + "-"
+        ssn["ssn"] += ssn.ssn_serial.astype(str).str.zfill(4)
+
+        return ssn["ssn"].to_numpy()
```

Comment:
Might just do `return (ssn_area.map(str) + "-" + ssn_group.map(str) + "-" + ssn_series.map(str)).to_numpy()`

Reply:
I'm using dot notation for the column access, which is also bad form.

Reply:
I don't see a logical difference between `.astype(str)` and `.map(str)` but [this](https://stackoverflow.com/questions/50782934/what-is-difference-between-mapstr-and-astypestr-in-dataframe) suggests `astype` is more performant than `map` for this case.

Reply:
I've updated this to be the one-line (pre-black) return for SSN and ITIN generation.